src/workers/webdav/modules/folders/infrastructure/HttpWebdavFolderRepository.ts

```diff
@@ -69,36 +66,45 @@ export class HttpWebdavFolderRepository implements WebdavFolderRepository {
       const folders = Object.entries(all).filter(([_, value]) => {
         if (!value.isFolder()) return false;
 
-        this.cleanOptimisticFolders(value);
-        return (
-          value.isFolder() &&
-          value.hasStatus(FolderStatuses.EXISTS) &&
-          !this.oldFolderIsBeingUpdated(value)
-        );
+        return value.isFolder() && value.hasStatus(FolderStatuses.EXISTS);
       }) as Array<[string, WebdavFolder]>;
 
-      const serverFolders = folders.reduce((items, [key, value]) => {
-        items[key] = value;
+      const serverFolders = folders.reduce((items, [path, value]) => {
+        this.cleanInMemoryFolderIfNeeded(path, value);
+
+        const existsInMemory =
+          this.inMemoryItems.exists(path) ||
+          this.inMemoryItems.existsByLastPath(path);
+        if (existsInMemory) {
+          return items;
+        }
+        items[path] = value;
 
         return items;
       }, {} as Record<string, WebdavFolder>);
 
-      const filteredOptimisticFolders: Record<string, WebdavFolder> = {};
+      const inMemoryFolders = this.inMemoryItems.getAllByType('FOLDER');
 
-      Object.keys(this.optimisticFolders).forEach((optimisticFolderPath) => {
-        if (
-          this.optimisticFolders[optimisticFolderPath].hasStatus(
-            FolderStatuses.EXISTS
-          )
-        ) {
-          filteredOptimisticFolders[optimisticFolderPath] =
-            this.optimisticFolders[optimisticFolderPath];
-        }
+      const filteredInMemoryFolders: Record<string, WebdavFolder> = {};
+
+      Object.keys(inMemoryFolders).forEach((path) => {
+        const inMemoryFolder = inMemoryFolders[path];
+        if (!inMemoryFolder.visible) return;
+
+        filteredInMemoryFolders[path] = WebdavFolder.from({
+          // -1 means this folder is being created and doesn't exist yet in the server
+          id: inMemoryFolder.externalMetadata?.id || -1,
```

Comment:
I'm against this, WebdavFolder, Files, etc are the core models and should always be valid ones. Not being able to trust everywhere else if a folder is valid or not will be a problem. For example, on every file action that requires its folder, we will need to check if the id is valid or not

Reply:
Do you have any suggestion on how to solve this?